### PR TITLE
[fix issue] check rg created successfully before deployment

### PR DESCRIPTION
**Pull request by LiliDeng**
Fix below issue
`
  File "D:\a\_work\1\s\lisa\lisa\runners\lisa_runner.py", line 276, in _deploy_environment_task
    self.platform.deploy_environment(environment)
  File "D:\a\_work\1\s\lisa\lisa\platform_.py", line 167, in deploy_environment
    self._deploy_environment(environment, log)
  File "D:\a\_work\1\s\lisa\lisa\sut_orchestrator\azure\platform_.py", line 492, in _deploy_environment
    raise identifier
  File "D:\a\_work\1\s\lisa\lisa\sut_orchestrator\azure\platform_.py", line 485, in _deploy_environment
    self._validate_template(deployment_parameters, log)
  File "D:\a\_work\1\s\lisa\lisa\sut_orchestrator\azure\platform_.py", line 1284, in _validate_template
    raise LisaException(error_message)
lisa.util.LisaException: ResourceGroupNotFound: Resource group 'lisa-Azure-Adhoc-20230117-024815-543-e145' could not be found.
`

### CODE REVIEW

File: lisa/sut_orchestrator/azure/common.py [hunk 1]
```diff
@@ -972,6 +972,16 @@ def check_or_create_resource_group(
                 rm_client.resource_groups.create_or_update(
                     resource_group_name, {"location": location}
                 )
+            timeout = 60
+            timeout_timer = create_timer()
+            while timeout_timer.elapsed(False) < timeout:
+                if rm_client.resource_groups.check_existence(resource_group_name):
+                    break
+                sleep(2)
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/common.py]: It's second time to use the timeout checking pattern, please create a method in utils/__init__.py.

Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/common.py]: updated

File: lisa/util/__init__.py [hunk 2]
```diff
@@ -657,3 +659,47 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
             value = value.strip() if isinstance(value, str) else value
             setattr(obj, field, value)
     return obj
+
+
+def check_get_attr(actual_object: Any, attr_name: str = "") -> Any:
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: `getattr_with_check`

File: lisa/util/__init__.py [hunk 3]
```diff
@@ -657,3 +659,47 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
...
+            current = actual_value
+    else:
+        actual_value = func(**test_kwargs)
+    if actual_value == expected:
+        return True
+    else:
+        return False
+
+
+def retry_till_timeout(func: Any, *func_args: Any, **kwargs: Any) -> None:
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: Try this example. if the logic is complicated, which is not able to be handled by lambda in one line, use a function.

```python
from functools import partial
from time import sleep
from typing import Callable

from lisa import LisaException
from lisa.util.logger import get_logger
from lisa.util.perf_timer import create_timer

log = get_logger()


def check_till_timeout(
    func: Callable[..., bool], timeout_message: str, timeout: int = 20
) -> None:
    timer = create_timer()
    while timer.elapsed(False) < timeout:
        if func():
            break
        sleep(1)
    if timer.elapsed() >= timeout:
        raise LisaException(f"timeout: {timeout_message}")


def method_with_arguments(wait_for: int = 5) -> bool:
    log.info("method_with_arguments")
    if global_wait_timer.elapsed(False) < wait_for:
        return False
    else:
        return True


global_wait_timer = create_timer()
log.info("calling check_till_timeout example 1...")
check_till_timeout(
    partial(method_with_arguments, wait_for=2),
    timeout_message="check_till_timeout example 1 is timeout",
)

global_wait_timer = create_timer()
log.info("calling check_till_timeout example 2...")
try:
    check_till_timeout(
        partial(method_with_arguments),
        timeout_message="check_till_timeout example 2 is timeout",
        timeout=2,
    )
except LisaException as identifier:
    log.info(identifier)

global_wait_timer = create_timer()
log.info("calling lambda 1...")
check_till_timeout(
    lambda: False if global_wait_timer.elapsed(False) < 2 else True,
    timeout_message="lambda example is timeout",
)

global_wait_timer = create_timer()
log.info("calling lambda 2...")

try:
    check_till_timeout(
        lambda: False if global_wait_timer.elapsed(False) < 5 else True,
        timeout_message="lambda example is timeout",
        timeout=2,
    )
except LisaException as identifier:
    log.info(identifier)

```

File: lisa/util/__init__.py [hunk 4]
```diff
@@ -657,3 +659,47 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
...
     return obj
+
+
+def check_get_attr(actual_object: Any, attr_name: str = "") -> Any:
+    if hasattr(actual_object, attr_name):
+        return getattr(actual_object, attr_name)
+    raise LisaException(f"attr {attr_name} not existing in object {actual_object}")
+
+
+def is_value_expected(
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: This method looks over engineered, it doesn't need.

File: lisa/util/__init__.py [hunk 5]
```diff
@@ -657,3 +659,38 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
...
+        raise LisaException(f"timeout: {timeout_message}")
+
+
+def getattr_with_check(actual_object: Any, attr_name: str = "") -> Any:
+    if hasattr(actual_object, attr_name):
+        return getattr(actual_object, attr_name)
+    raise LisaException(f"attr {attr_name} not existing in object {actual_object}")
+
+
+def method_with_arguments(
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: method_with_arguments just an example, it doesn't need. Try to use the logic by lambda directly. If the logic is multiple lines, wrapper it to a function.

File: lisa/sut_orchestrator/azure/common.py [hunk 6]
```diff
@@ -1042,18 +1050,17 @@ def wait_copy_blob(
...
-            break
-        # the copy is very slow, it may need several minutes. check it every
-        # 2 seconds.
-        sleep(2)
-    if timeout_timer.elapsed() >= timeout:
-        raise LisaException(f"wait copying VHD timeout: {vhd_path}")
-
+    check_till_timeout(
+        partial(
+            method_with_arguments,
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/common.py]: Try this.

```python
check_till_timeout(lambda: blob_client.get_blob_properties().copy.status == "success", timeout_message...
```

Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/common.py]: thanks, this works.

File: lisa/util/__init__.py [hunk 7]
```diff
@@ -657,3 +659,18 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
             value = value.strip() if isinstance(value, str) else value
             setattr(obj, field, value)
     return obj
+
+
+def check_till_timeout(
+    func: Callable[..., Any],
+    timeout_message: str,
+    timeout: int = 20,
+    waittime: int = 1,
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: usually, it's called `interval`

File: lisa/util/__init__.py [hunk 8]
```diff
@@ -657,3 +659,18 @@ def strip_strs(obj: Any, fields: List[str]) -> Any:
             value = value.strip() if isinstance(value, str) else value
             setattr(obj, field, value)
     return obj
+
+
+def check_till_timeout(
+    func: Callable[..., Any],
+    timeout_message: str,
+    timeout: int = 20,
```
Comment (squirrelsc) [inline lisa/util/__init__.py]: Give 60 as the default value, so at least one call doesn't need to specify it.

File: lisa/util/__init__.py (hunk 7 above)
Comment (LiliDeng) [reply lisa/util/__init__.py]: updated

File: lisa/util/__init__.py (hunk 8 above)
Comment (LiliDeng) [reply lisa/util/__init__.py]: updated

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**